Break RefrenceFrame.orient() into seperate methods.

sidhu1012 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
Fixes #17758 
Fixes #18798
#### Brief description of what is fixed or changed
Break RefrenceFrame.orient() into seperate methods.

`.orient()` is trying to do too many things, so this change both organizes the library code better as well as provide users with easier to understand and use methods. We do not plan to deprecate `.orient()` as it would disrupt too much code in the wild.

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
- physics.vector
      - Introduced new methods on ReferenceFrame: `.orient_axis()`, `.orient_explicit()`, `.orient_body_fixed()`, `.orient_space_fixed()`, and `.orient_quaternion()`. `.orient()` calls out to each new method and it is recommended to use the new methods. 
<!-- END RELEASE NOTES -->

sksingh1202 committed:
Rewritten ReferenceFrame.orient() into 5 separate methods and modified the methods which use ReferenceFrame.orient()

sidhu1012 committed:
Continued work on breaking up .orient().

- Updated prior work to work with latest sympy.
- Added test cases for every orient function.
- Added a new helper function to reduce duplicate code.

oscarbenjamin commented:
This PR removes all of the blame history for the `test_solvers.py` file because the file is deleted and then added again.

It would be better to squash the commits into a single commit with @sksingh1202 listed as co-author.

sidhu1012 commented:
@moorepants Please review and tell what's need to be done further.

sidhu1012 commented:
@moorepants I have added test cases , added back orient function, fixed docstring abd everything is working
Test failed because of HTML error.
 
What need's to be done further

sidhu1012 commented:
@moorepants Please review

sidhu1012 commented:
Is this an error with code or build?

Makefile:49: recipe for target 'html' failed
make: *** [html] Error 1

sidhu1012 commented:
@moorepants Finally passed all tests. Please review now.

moorepants commented:
I'll try to get to it soon. Thanks.

sidhu1012 commented:
Closing and opening to restart tests.

sidhu1012 commented:
@moorepants Please Review.

czgdp1807 commented:
@moorepants This seems ready. Are there any more changes required to conclude this PR. Thank you.

moorepants commented on sympy/physics/vector/frame.py:
```diff
@@ -476,71 +476,63 @@ def dcm(self, otherframe):
+        self._dcm_cache.update({parent: parent_orient.T})
+        parent._dcm_cache.update({self: parent_orient})
+
+    def orient_axis(self, parent, amounts):
```
```suggestion
    def orient_axis(self, parent, axis, amount):
```

Rotation about an axis requires 3 three things: 1) parent frame to rotate relative to, 2) vector (axis) to rotate about, and 3) a scalar amount of rotation.

The api should be:

```
def orient_axis(self, reference_frame, vector, amount):
```

or

```
def orient_axis(self, parent_frame, vector_axis, amount):
```

It would be nice if the argument variable names were informative.

moorepants commented on sympy/physics/vector/frame.py:
```diff
@@ -476,71 +476,63 @@ def dcm(self, otherframe):
+
+    def orient_axis(self, parent, amounts):
+        """Sets the orientation of this frame with respect to a another(parent)
+        reference frame by rotating about a fixed axis.
```
I'm not sure but this might mean "...by rotating about an axis fixed in the parent reference frame." or maybe that isn't the restriction. We should clarify what the axis can and can't be, because it is very easy to pass in a vector that is varying wrt time.

moorepants commented on sympy/physics/vector/frame.py:
```diff
@@ -476,71 +476,63 @@ def dcm(self, otherframe):
+        amounts : 2-tuple (expr/sym/func, Vector)
+            The first element is the amount by which it the frame is to be rotated.
+            The second element is the axis(a vector in the parent frame) about which
+            the frame is to be rotated. It need not be a unit vector.
```
This should be broken into two arguments.

moorepants commented on sympy/physics/vector/frame.py:
```diff
@@ -554,45 +546,145 @@ def orient(self, parent, rot_type, amounts, rot_order=''):
+        parent_orient_axis = []
+
+        if not axis.dt(parent) == 0:
+            raise ValueError('Axis cannot be time-varying')
```
So this check ensures the axis cannot be moving in the parent frame. So the language can be updated to be more clear in the docstring.

moorepants commented on sympy/physics/vector/frame.py:
```diff
@@ -554,45 +546,145 @@ def orient(self, parent, rot_type, amounts, rot_order=''):
+        parent._ang_vel_dict.update({self: -wvec})
+        self._var_dict = {}
+
+    def orient_dcm(self, parent, amounts):
```
This could likely have a better function name. Some ideas:

- orient_explicit (because we are explicit defining the direction cosine matrix (rotation matrix))
- orient_direction_cosine_matrix
- orient_from_direction_cosine_matrix
- orient_rotation_matrix
- orient_from_rotation_matrix

moorepants commented on sympy/physics/vector/frame.py:
```diff
@@ -554,45 +546,145 @@ def orient(self, parent, rot_type, amounts, rot_order=''):
+                           [sin(angle), cos(angle), 0],
+                           [0, 0, 1]])
+
+    def orient_body(self, parent, amounts, rot_order=''):
```
`orient_body_fixed` is a bit more informative.

moorepants replied:
`rot_order` should not be an optional kwarg.

moorepants commented on sympy/physics/vector/frame.py:
```diff
@@ -616,18 +708,90 @@ def orient(self, parent, rot_type, amounts, rot_order=''):
+        parent._ang_vel_dict.update({self: -wvec})
+        self._var_dict = {}
+
+    def orient_space(self, parent, amounts, rot_order=''):
```
`orient_space_fixed`

rot_order is not optional

sidhu1012 commented:
@moorepants Done! Ready for review 😂

moorepants commented on sympy/physics/vector/frame.py:
```diff
@@ -554,45 +549,140 @@ def orient(self, parent, rot_type, amounts, rot_order=''):
+        parent_frame._ang_vel_dict.update({self: -wvec})
+        self._var_dict = {}
+
+    def orient_explicit(self, parent, amounts):
```
```suggestion
    def orient_explicit(self, parent, dcm):
```

moorepants commented on sympy/physics/vector/frame.py:
```diff
@@ -554,45 +549,140 @@ def orient(self, parent, rot_type, amounts, rot_order=''):
 
-        The axis does not have to be defined by a unit vector, it can be any
-        vector in the parent frame.
+        amounts : Matrix, shape(3, 3)
```
```suggestion
        dcm : Matrix, shape(3, 3)
```

moorepants commented on sympy/physics/vector/frame.py:
```diff
@@ -554,45 +549,140 @@ def orient(self, parent, rot_type, amounts, rot_order=''):
+
+    def orient_explicit(self, parent, amounts):
+        """Sets the orientation of this reference frame relative to another
+        (parent) reference frame by setting the direction cosine matrix directly.
```
```suggestion
        (parent) reference frame by setting the direction cosine matrix explicitly.
```

moorepants commented on sympy/physics/vector/frame.py:
```diff
@@ -554,45 +549,140 @@ def orient(self, parent, rot_type, amounts, rot_order=''):
+            to.
+
+        amounts : 3-tuple of expressions, symbols, or functions
+        The string ``'123'`` and integer ``123`` are equivalent.
```
```suggestion
             The string ``'123'`` and integer ``123`` are equivalent.
```

sidhu1012 commented on sympy/physics/vector/frame.py:
```diff
@@ -690,154 +918,114 @@ def orient(self, parent, rot_type, amounts, rot_order=''):
 
-    def orientnew(self, newname, rot_type, amounts, rot_order='',
-                  variables=None, indices=None, latexs=None):
+    def orientnew(self, newname, rot_type, amounts, rot_order='', variables=None, indices=None, latexs=None):
```
I'm not sure what needs to be done here.

sidhu1012 replied:
Still did what you I thought you mean.

sidhu1012 commented:
@moorepants Please review.

moorepants committed:
Improvements to the docstrings.

moorepants commented:
@sidhu1012 I added a commit with some improvements to the new docstrings. Have a look to see if you agree with the changes.

sidhu1012 committed:
Fix axis orientation

sidhu1012 commented:
> @sidhu1012 I added a commit with some improvements to the new docstrings. Have a look to see if you agree with the changes.

New docstring looks much better. I have fixed the failing tests. Does it look good now?

sidhu1012 committed:
Fix space and explicit orientation docstring

sidhu1012 committed:
Fix import in docstring

sidhu1012 committed:
Update explicit docstring

moorepants committed:
Final adjustment on docstrings and variable names.

moorepants committed:
Fixed math notation error, added note about new methods, and link to quaternion article.

moorepants commented:
Ok, I'm good with this now. If tests pass it can be merged.

moorepants commented:
We need to do this:

> This PR removes all of the blame history for the test_solvers.py file because the file is deleted and then added again.

> It would be better to squash the commits into a single commit with @sksingh1202 listed as co-author.

I'm not sure how to squash and retain contributions from the three authors though.

oscarbenjamin commented:
> I'm not sure how to squash and retain contributions from the three authors though.

I would just squash to a single commit and use Co-authored-by

moorepants commented:
I assume you mean to simply mention the authors in the commit message?

oscarbenjamin commented:
There is an established convention to use `Co-authored-by`:
https://stackoverflow.com/questions/7442112/how-to-attribute-a-single-commit-to-multiple-developers

This is what I do if I squash the commits for someone else e.g.: https://github.com/sympy/sympy/pull/20771

moorepants commented:
I dropped some of the offending commits and squashed some, so hopefully that will work and preserve some history of who added things.

moorepants commented:
Ok, seems this is good to go.

oscarbenjamin commented:
This PR seems to have broken the doctests somehow.

The failure that I am seeing on master is:
```
$ bin/doctest sympy/physics/vector/frame.py 
====================================================== test process starts =======================================================
executable:         /Users/enojb/current/sympy/39venv/bin/python  (3.9.0-final-0) [CPython]
architecture:       64-bit
cache:              yes
ground types:       python 
numpy:              1.20.1
hash randomization: on (PYTHONHASHSEED=1142599648)

sympy/physics/vector/frame.py[15] .....F.........                                                                           [FAIL]

__________________________________________________________________________________________________________________________________
__________________________________ sympy.physics.vector.frame.ReferenceFrame.orient_space_fixed __________________________________
File "/Users/enojb/current/sympy/sympy/sympy/physics/vector/frame.py", line 838, in sympy.physics.vector.frame.ReferenceFrame.orient_space_fixed
Failed example:
    B.dcm(N).simplify()
Expected:
    Matrix([
    [ sin(q1)*sin(q2)*sin(q3) + cos(q1)*cos(q3), sin(q1)*cos(q2),                                                                                sin(q1)*sin(q2)*cos(q3) - sin(q3)*cos(q1)],
    [-sin(q1)*cos(q3) + sin(q2)*sin(q3)*cos(q1), cos(q1)*cos(q2), sin(-q1 + q2 + q3)/4 - sin(q1 - q2 + q3)/4 + sin(q1 + q2 - q3)/4 + sin(q1 + q2 + q3)/4 + cos(q1 - q3)/2 - cos(q1 + q3)/2],
    [                           sin(q3)*cos(q2),        -sin(q2),                                                                                                          cos(q2)*cos(q3)]])
Got:
    Matrix([
    [ sin(q1)*sin(q2)*sin(q3) + cos(q1)*cos(q3), sin(q1)*cos(q2), sin(q1)*sin(q2)*cos(q3) - sin(q3)*cos(q1)],
    [-sin(q1)*cos(q3) + sin(q2)*sin(q3)*cos(q1), cos(q1)*cos(q2), sin(q1)*sin(q3) + sin(q2)*cos(q1)*cos(q3)],
    [                           sin(q3)*cos(q2),        -sin(q2),                           cos(q2)*cos(q3)]])

===================================== tests finished: 14 passed, 1 failed, in 33.69 seconds ======================================
DO *NOT* COMMIT!
```

I've bisected to one of these three commits:

21e1f3f9dd9caf515277838060745b76bb3fe012
91af2e8b801546d526019ff663b52322b4db7e40
ef876cd9b0b25988210c00e4ed7c06fb4cca7d53

I can't bisect more precisely because there were many completely broken commits with missing imports etc.

sidhu1012 commented:
@oscarbenjamin I'll look into it.

moorepants commented:
Interesting. It passed locally and the PR CI passed. Weird that it would have changed.

I think all this is, is that simplify() isn't bringing the expression to the expected (acutally simpler) form.

@sidhu1012 If you are going to fix this, it would be nice if it did reproduce the expected transpose of the `orient_body_fixed()` method. I'm not sure why it wouldn't.

I also notice that this doctest was skipped when I was editing and I disabled the skip, tested, and it seemed to work.

moorepants commented:
The single entry in the matrix should simplify better. Maybe a different method in trigsimp is needed.

moorepants commented:
Actually the second thing is the correct thing. The expected value in the docstring is in correct. Not sure where that came from.

sidhu1012 commented:
> Actually the second thing is the correct thing. The expected value in the docstring is in correct. Not sure where that came from.

Yes I also noticed that. I'll fix it tomorrow.

moorepants commented:
I just fixed it. Making PR now.

oscarbenjamin commented:
> Interesting. It passed locally and the PR CI passed. Weird that it would have changed.

I'm not sure what happened there. We have a few problems in CI right now that I'm trying to fix.
